### ISSUE TITLE
support for iterables and iterators (in general)

### ISSUE THREAD

**Issue by jmagaram** [ramda/ramda#2525]
My background is C# and I love LINQ. I also have had great experiences with F#. So what's the best alternative for TypeScript? First I tried some npm modules like 'linq' and 'linq-es2015'. I prefer the dot-less more functional approach and have run into a basic problems, like the GroupBy method exploding. I couldn't get my head around underscore at all. I thought Ramda might be the answer. But it doesn't seem to support iterables - only arrays. See the example code below. I would expect ALL the LINQ-like and F# seq operators - groupBy, scan, minBy, orderBy, zip, pairwise, etc. to be supported out of the box, but no. It would also be helpful to have some basic generators, like R.range(1,100). Am I wrong or is this whole class of functionality a natural fit for Ramda? Is so, why isn't it there yet? LINQ is a critical technology for nearly all C# .net development and it has been around for years. Can't imagine doing .net development without it. But in this JavaScript world I can't find it even though modern browsers with ES6 support it. [f# seq module](https://msdn.microsoft.com/visualfsharpdocs/conceptual/collections.seq-module-%5bfsharp%5d) and [.net linq](https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable?view=netframework-4.7.2)

    test("use filter on an array", () => {
        let result = R.filter(isEven, [1, 2, 3, 4]);
        expect(result).toEqual([2, 4]);
    });

    function* numbers() {
        yield 1;
        yield 2;
        yield 3;
        yield 4;
    }

    test("use filter on an iterable", () => {
        // no lazy evaluation here. can't get this to compile without first
        // materializing the iterable into an array.
        let result = R.filter(isEven, Array.from(numbers()));
        expect(result).toEqual([2, 4]);
    });

**Comment by CrossEye** [comment]
> Am I wrong or is this whole class of functionality a natural fit for Ramda? 

No, you're not wrong at all.  There is good reason to want these integrations.  There was some recent discussion in #2508.  But nothing has happened yet.  This would be a large change, with a lot of code.  Also, while @semmel made it clear enough in that issue how we would be able to support iterators, I'm less sure about generators.

You ask why this is not in Ramda so far.  That's because it has never bubbled to the top of the concerns for any of the core team.

There are also questions about exactly *what* functions to support.  `reduce`, for instance`, makes perfect sense on a list or finite iterable, we don't know ahead of time that an iterable is finite.  So do we just implement it and let it (slowly) fail on an infinite one?  Or do we not implement it at all for iterables?

Overally, it's a good idea, and a close fit for Ramda.  But it is a lot of work.  Are you interested in pitching in, working on a few functions?

----------

You should know that LINQ is likely not a motivator for most of the core team, nor is C#.  The inspirations for the library come from the two opposite ends of the functional programming world: from Haskell, with a smattering of Scala and from LISPy languages such as Scheme and Clojure.  The most proximate cause for the library's inception was the publication of @raganwald's [_JavaScript Allongé_][ja].  It's not that we don't appreciate F# or OCaml, or Erlang or other FP languages, but I think the main motivations are from the Haskell and LISP worlds.  So your enthusiasm for LINQ doesn't translate here to any particular actions; it's just not what makes us tick.

Meanwhile, while some people have written Ramda typings for Typescript and Flow, Ramda itself tends to be concerned only with Javascript.

----------

There are a number of libraries that focus on these concerns, if trying to add them to Ramda seems overwhelming.  Probably the most popular is [wu.js][wu], although I'm fond of [ES-Iter][ei], and there are many more:

  - https://github.com/blond/ho-iter
  - https://github.com/ominibyte/richflow
  - https://github.com/StephenCleary/iterjs
  - https://www.npmjs.com/package/rainboom
  - https://github.com/sulcata/reiter
  - https://github.com/rfw/sloth.js
  - https://github.com/benji6/imlazy
  - https://github.com/xgbuils/iterum
  - https://github.com/DanShappir/Seq/
  - https://github.com/zeusdeux/lazit/
  - https://github.com/seangenabe/starry



  [ja]: https://leanpub.com/javascript-allonge
  [wu]: https://github.com/fitzgen/wu.js/
  [ei]: https://github.com/abozhilov/ES-Iter

**Comment by jmagaram** [comment]
I started learning Javascript and Typescript 2 weeks ago so I’m not very qualified. I can imagine writing some functions from scratch in Typescript, but integrating with Ramda would be a big learning curve – just started exploring it today. I also can’t handle programming without static types – Typescript is essential for me and the Ramda code is all regular js. F# is a great language and that Seq package is a great model for what functions are important. I don't think it would take long for someone with more experience to build it. These little Typescript functions below seem to work. I'll check out the libraries you mentioned - thanks.

    export function* filter<T>(source: Iterable<T>, predicate: (item: T) => boolean) {
        for (const i of source) {
            if (predicate(i)) {
                yield i;
            }
        }
    }

    export function* map<T, U>(source: Iterable<T>, selector: (t: T) => U) {
        for (const i of source) {
            yield selector(i);
        }
    }

**Comment by CrossEye** [comment]
> I started learning Javascript and Typescript 2 weeks ago so I’m not very qualified.

We should never let that stop us! :smile: 

While these functions look fine and could be turned into plain JS simply by removing the typings in the signatures, the real problem would be how they integrate with the remainder of the signatures.  `map`, for instance, currently works with functions, objects, and arrays, through Ramda code, but with any object that has either a `map` method or a `fantasyland/map` method through dispatching.  We could presumably add arbitrary Iterables in this way, but it would have to come after the others, I would think, because we would want `filter(n => n > 5, [2, 3, 5, 7, 11])` to return `[7, 11]`, not `{[Symbol.iterator]: someFn}`.  So it would take some careful work to integrate such.

And this still doesn't address question of whether passing an infinite iterable to `reduce` should simply be treated as Garbage In/Garbage Out or if we shouldn't convert `reduce`.

There are some thorny questions here.  I think we can work them out, but it's a difficult process.

----------

> I also can’t handle programming without static types – Typescript is essential for me 

If you find static types so useful, you might want to check out Haskell or its compile-to-JS variant, PureScript.  These have much more powerful typings than something like Java or C#.  F# is a cousin to these; they all descend from ML.

**Comment by Izhaki** [comment]
Having said that (@CrossEye) it's +1 for me for iterators, with `R.chain` being the prime use case for me.

**Comment by jkoudys** [comment]
@CrossEye Hey Scott, I'm very experienced and happy to help out with this.

My main goal is to let me use Ramda to work with collections aside from arrays and pojos. e.g. the difference in lodash's map vs Ramda sets a good precedent for other types, since `{ a: 1, b: 2 } |> map(v => v * 2)` gives `[2, 4]` in Lo but `{ a: 2, b: 4 }` in Ramda. To me that suggests Ramda likes to keep your original collection type, while lodash wants everything to be an array, which I don't find super useful since I can `[...myCollection]` anyways.

What I'd love to implement is something so I can use Map and Set immutably, and with all those methods we take for granted in the Array.prototype. `new Set([1, 2]) |> map(v => v * 2)` returning a Set with 2, 4 would be pretty awesome.

The question of generators (or any unbound iterable) is an interesting one. Frankly those seem like a much better fit for a project like rxjs, which is built to nicely wrap the iterator so it could continue (observing next()s over time) or eventually complete(). If it's unbound, I'd assume I'd see the same behaviour as if I called `Array.from((function* () { while(true) yield 1; })())`, which is to run forever.

I see https://github.com/ramda/ramda/pull/2585 has in the next-steps some mention of Map, so I'll take a look on that PR and see what I can come up with.

**Comment by CrossEye** [comment]
Hello, Joshua, and welcome!

I would love to see what you come up with.

I've seriously considered adding `R.Map` and `R.Set` types, even though Ramda has shied away from creating types.  The problem is that the built-in `Map` and `Set` types are based on reference equality, and Ramda is all about value equality.

That said, some of our functions could clearly be extended to the built-in `Map` and `Set` types.

> [T]he difference in lodash's map vs Ramda sets a good precedent for other types, since `{ a: 1, b: 2 } |> map(v => v * 2)` gives `[2, 4]` in Lo but `{ a: 2, b: 4 }` in Ramda. To me that suggests Ramda likes to keep your original collection type, while lodash wants everything to be an array, which I don't find super useful since I can `[...myCollection]` anyways.

It's more than just that Ramda "likes to" do so.  Ramda tries hard to work with the [FantasyLand Specification][fl], and the rules for the [`Functor`][fu] type (which specifies `map`) require that 
`map(fn, myFunctor)` return a Functor of the same type as `myFunctor`.  Changing from Object to Array is simply not allowed.

I look forward to seeing your work.  Please let us know how we can help.



  [fl]: https://github.com/fantasyland/fantasy-land
  [fu]: https://github.com/fantasyland/fantasy-land#functor
